flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -111,6 +108,14 @@
 	/** The executor for potentially blocking I/O operations, like state disposal. */
 	private final Executor executor;
 
+	/** The executor for non-blocking operations. */
+	private final Executor mainThreadExecutor;
+
+	private final CompletedCheckpointStore completedCheckpointStore;
+
+	/** The lock for avoiding conflict between I/O operations. */
+	private final Object operationLock = new Object();
```

Comment:
Yes, there is a small possibility that the `CheckpointCoordinator` is shut down when a `PendingCheckpoint` is doing finalization. There could be some concurrent conflicts on `operatorStates` and `targetLocation`. It might be not a big deal because it would be shut down anyway. The finalization probably could not finish because the IO executor would be also shut down. However it's not so elegant to leave the concurrent issue to the `CheckpointStorageLocation` and `OperatorState`. And it's a bit heavy to make all of these implementations thread-safe to avoid the small possibility issue. So here I think introducing a lock outside is better.